t/integration/test_canvas.py

```diff
@@ -452,6 +453,138 @@ def test_nested_chain_group_last(self, manager):
         res = sig.delay()
         assert res.get(timeout=TIMEOUT) == [42, 42]
 
+    def test_chain_replaced_with_a_chain_and_a_callback(self, manager):
+        if not manager.app.conf.result_backend.startswith('redis'):
+            raise pytest.skip('Requires redis result backend.')
+
+        redis_connection = get_redis_connection()
+        redis_connection.delete('redis-echo')
+
+        link_msg = 'Internal chain callback'
+        c = chain(
+            identity.s('Hello '),
+            # The replacement chain will pass its args though
+            replace_with_chain.s(link_msg=link_msg),
+            add.s('world'),
+        )
+        res = c.delay()
+
+        assert res.get(timeout=TIMEOUT) == 'Hello world'
+
+        expected_msgs = {link_msg, }
+        while expected_msgs:
+            maybe_key_msg = redis_connection.blpop('redis-echo', TIMEOUT)
+            if maybe_key_msg is None:
+                raise TimeoutError('redis-echo')
+            _, msg = maybe_key_msg
+            msg = msg.decode()
+            expected_msgs.remove(msg)   # KeyError if `msg` is not in here
+
+        # There should be no more elements - block momentarily
+        assert redis_connection.blpop('redis-echo', min(1, TIMEOUT)) is None
+        redis_connection.delete('redis-echo')
+
+    def test_chain_replaced_with_a_chain_and_an_error_callback(self, manager):
+        if not manager.app.conf.result_backend.startswith('redis'):
+            raise pytest.skip('Requires redis result backend.')
+
+        redis_connection = get_redis_connection()
+        redis_connection.delete('redis-echo')
+
+        link_msg = 'Internal chain errback'
+        c = chain(
+            identity.s('Hello '),
+            replace_with_chain_which_raises.s(link_msg=link_msg),
+            add.s(' will never be seen :(')
+        )
+        res = c.delay()
+
+        with pytest.raises(ValueError):
+            res.get(timeout=TIMEOUT)
+
+        expected_msgs = {link_msg, }
```

Comment:
only {link_msg, } ?

Reply:
Yep, I left it like this so that we could refactor these tests with paramaterize and maybe subtest on a future pass through the file. If they all expect to check against a set of messages we should be able to handle that in a parameterized test a bit easier.